model_softnms.py

```diff
@@ -0,0 +1,2559 @@
+        rpn_bbox: [batch, anchors, (dy, dx, log(dh), log(dw))]
+
+    Returns:
+        Proposals in normalized coordinates [batch, rois, (y1, x1, y2, x2)]
+    """
+
+    def __init__(self, proposal_count, nms_threshold, anchors,
+                 config=None, **kwargs):
+        """
+        anchors: [N, (y1, x1, y2, x2)] anchors defined in image coordinates
+        """
+        super(ProposalLayer, self).__init__(**kwargs)
+        self.config = config
+        self.proposal_count = proposal_count
+        self.nms_threshold = nms_threshold
+        self.anchors = anchors.astype(np.float32)
+
+    def call(self, inputs):
+        # Box Scores. Use the foreground class confidence. [Batch, num_rois, 1]
+        scores = inputs[0][:, :, 1]
+        # Box deltas [batch, num_rois, 4]
+        deltas = inputs[1]
+        deltas = deltas * np.reshape(self.config.RPN_BBOX_STD_DEV, [1, 1, 4])
+        # Base anchors
+        anchors = self.anchors
+
+        # Improve performance by trimming to top anchors by score
+        # and doing the rest on the smaller subset.
+        pre_nms_limit = min(6000, self.anchors.shape[0])
+        ix = tf.nn.top_k(scores, pre_nms_limit, sorted=True,
+                         name="top_anchors").indices
+        scores = utils.batch_slice([scores, ix], lambda x, y: tf.gather(x, y),
+                                   self.config.IMAGES_PER_GPU)
+        deltas = utils.batch_slice([deltas, ix], lambda x, y: tf.gather(x, y),
+                                   self.config.IMAGES_PER_GPU)
+        anchors = utils.batch_slice(ix, lambda x: tf.gather(anchors, x),
+                                    self.config.IMAGES_PER_GPU,
+                                    names=["pre_nms_anchors"])
+
+        # Apply deltas to anchors to get refined anchors.
+        # [batch, N, (y1, x1, y2, x2)]
+        boxes = utils.batch_slice([anchors, deltas],
+                                  lambda x, y: apply_box_deltas_graph(x, y),
+                                  self.config.IMAGES_PER_GPU,
+                                  names=["refined_anchors"])
+
+        # Clip to image boundaries. [batch, N, (y1, x1, y2, x2)]
+        height, width = self.config.IMAGE_SHAPE[:2]
+        window = np.array([0, 0, height, width]).astype(np.float32)
+        boxes = utils.batch_slice(boxes,
+                                  lambda x: clip_boxes_graph(x, window),
+                                  self.config.IMAGES_PER_GPU,
+                                  names=["refined_anchors_clipped"])
+
+        # Filter out small boxes
+        # According to Xinlei Chen's paper, this reduces detection accuracy
+        # for small objects, so we're skipping it.
+
+        # Normalize dimensions to range of 0 to 1.
+        normalized_boxes = boxes / np.array([[height, width, height, width]])
+
+        # Non-max suppression
+        def nms(normalized_boxes, scores):
+            indices, scores = tf.image.non_max_suppression_with_scores(
```

Comment:
This matches the call I made when generating the initial "softmax" weights, though running detection with the corresponding weights does not yield any results.